Sources/Plasma/CoreLib/hsBitVector.cpp

```diff
@@ -50,31 +50,33 @@ void hsBitVector::IGrow(uint32_t newNumBitVectors)
     hsAssert(newNumBitVectors > fNumBitVectors, "Growing smaller");
     uint32_t *old = fBitVectors;
     fBitVectors = new uint32_t[newNumBitVectors];
-    int i;
-    for( i = 0; i < fNumBitVectors; i++ )
+
+    uint32_t i;
+    for (i = 0; i < fNumBitVectors; i++)
         fBitVectors[i] = old[i];
-    for( ; i < newNumBitVectors; i++ )
+    for (; i < newNumBitVectors; i++)
         fBitVectors[i] = 0;
-    delete [] old;
+    delete[] old;
     fNumBitVectors = newNumBitVectors;
 }
 
 hsBitVector& hsBitVector::Compact()
 {
-    if( !fBitVectors )
+    if (!fBitVectors)
         return *this;
 
-    if( fBitVectors[fNumBitVectors-1] )
+    if (fBitVectors[fNumBitVectors-1])
         return *this;
 
-    int hiVec = 0;
-    for( hiVec = fNumBitVectors-1; (hiVec >= 0)&& !fBitVectors[hiVec]; --hiVec );
-    if( hiVec >= 0 )
+    uint32_t hiVec = 0;
+    for (hiVec = fNumBitVectors-1; (int32_t(hiVec) >= 0) && !fBitVectors[hiVec]; --hiVec)
+        ;
+
+    if (hiVec >= 0)
```

Comment:
`hiVec` should be `int32_t` -- despite initializing it with a `uint32_t - 1`, it is compared with 0 in two places (the latter without a cast, which will always be `true`)

Reply:
Thinking about this more, the core issue seems to be that limits defined/enforced for what the maximum value in a bit vector is. At one point in `Enumerate`, it tries to store all the contained bits in a `vector<uint16_t>`. It's clearly possible to store bits that exceed the capacity of a 16-bit unsigned integer, and even that exceed a 32-bit signed integer.

Should we start enforcing that we only support storing up to `uint16_t` max value, which would make it entirely safe to use a `int32_t` for the number of bit vectors and allow us to use -1 with no jumping between signedness?

Reply:
Worse, `Enumerate` populates a vector of `int16_t`, even though it will never use negative indices...  I'm fine with enforcing a maximum size of 2^15 bits, but I don't necessarily feel that `fNumBitVectors` needs to become signed.  For reference, I already use `hsSsize_t` for backwards iteration in several places with `std::vector`s, with the implicit assumption that those vectors will never be greater than 2^31 (or 2^63 on 64-bit) elements.  If we have that many of anything in Plasma, Something Terribly has Gone Wrong ;)